firmware/user/modes/mode_tunernome.c

```diff
@@ -0,0 +1,721 @@
+
+#include "buttons.h"
+#include "user_main.h"
+#include "embeddednf.h"
+#include "oled.h"
+#include "bresenham.h"
+#include "assets.h"
+#include "buttons.h"
+#include "linked_list.h"
+#include "font.h"
+#include "mode_colorchord.h"
+
+#include "embeddednf.h"
+#include "embeddedout.h"
+
+/*============================================================================
+ * Defines, Structs, Enums
+ *==========================================================================*/
+
+#ifndef M_PI
+#define M_PI		3.14159265358979323846
+#endif
+
+#define TUNERNOME_UPDATE_MS 20
+#define TUNERNOME_UPDATE_S (TUNERNOME_UPDATE_MS / 1000.0f)
+
+#define NUM_STRINGS            6
+#define GUITAR_OFFSET          0
+#define CHROMATIC_OFFSET       6 // adjust start point by quartertones   
+#define SENSITIVITY            5
+
+#define METRONOME_CENTER_X OLED_WIDTH / 2
+#define METRONOME_CENTER_Y OLED_HEIGHT - 10
+#define METRONOME_RADIUS 35
+#define INITIAL_BPM 60
+#define MAX_BPM 600
+#define METRONOME_FLASH_MS 50
+
+/// Helper macro to return an integer clamped within a range (MIN to MAX)
+#define CLAMP(X, MIN, MAX) ( ((X) > (MAX)) ? (MAX) : ( ((X) < (MIN)) ? (MIN) : (X)) )
+/// Helper macro to return the absolute value of an integer
+#define ABS(X) (((X) < 0) ? -(X) : (X))
+
+typedef enum
+{
+    TN_TUNER,
+    TN_METRONOME
+} tnMode;
+
+typedef enum
+{
+    GUITAR_TUNER = 0,
+    SEMITONE_0,
+    SEMITONE_1,
+    SEMITONE_2,
+    SEMITONE_3,
+    SEMITONE_4,
+    SEMITONE_5,
+    SEMITONE_6,
+    SEMITONE_7,
+    SEMITONE_8,
+    SEMITONE_9,
+    SEMITONE_10,
+    SEMITONE_11,
+    MAX_GUITAR_MODES
+} tuner_mode_t;
+
+typedef struct
+{
+    tnMode mode;
+    tuner_mode_t currentMode;
+
+    timer_t updateTimer;
+    timer_t ledTimer;
+
+    int audioSamplesProcessed;
+    uint32_t intensities_filt[NUM_STRINGS];
+    int32_t diffs_filt[NUM_STRINGS];
+
+    int frame;
+    int pause;
+    int lastX, lastY;
+    int bpm;
+
+    //mostly temp values
+    float bps;
+    float periodS;
+    float periodMS;
+    float cycleFrames;
+    float halfCycleFrames;
+    int tockFrame1;
+    int tockFrame2;
+    int finalCycleFrame;
+
+    uint32_t semitone_intensitiy_filt;
+    int32_t semitone_diff_filt;
+} tunernome_t;
+
+/*============================================================================
+ * Prototypes
+ *==========================================================================*/
+
+void ICACHE_FLASH_ATTR tunernomeEnterMode(void);
+void ICACHE_FLASH_ATTR tunernomeExitMode(void);
+void ICACHE_FLASH_ATTR switchToSubmode(tnMode);
+void ICACHE_FLASH_ATTR tunernomeButtonCallback(uint8_t state __attribute__((unused)),
+        int button, int down);
+void ICACHE_FLASH_ATTR tunernomeSampleHandler(int32_t samp);
+void ICACHE_FLASH_ATTR recalcMetronome();
+static void ICACHE_FLASH_ATTR tunernomeUpdate(void* arg __attribute__((unused)));
+void ICACHE_FLASH_ATTR ledReset(void* timer_arg __attribute__((unused)));
+
+static inline int16_t getMagnitude(uint16_t idx);
```

Comment:
Note to self, these are `inline`, but make sure they're not getting linked in iram.

Reply:
These functions don't show up in `image.map` at all, truly inlined.